From: Paul Gortmaker <paul.gortmaker@windriver.com>

hrtimer: spin on hardirq context callback

The hrtimers-stuck-in-waitqueue.patch indicated that timers which
run their callback from hardirq context should be waited for.

However, the HRTIMER_CB_IRQSAFE_NO_SOFTIRQ categorization that was
used in 2.6.26 is no longer, due to commit ccc7dad, which introduces
HRTIMER_CB_IRQSAFE_PERCPU and HRTIMER_CB_IRQSAFE_UNLOCKED.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 kernel/hrtimer.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/kernel/hrtimer.c b/kernel/hrtimer.c
index 8093447..1eee1a4 100644
--- a/kernel/hrtimer.c
+++ b/kernel/hrtimer.c
@@ -1065,8 +1065,9 @@ int hrtimer_cancel(struct hrtimer *timer)
 		if (ret >= 0)
 			return ret;
 		switch (timer->cb_mode) {
-		case HRTIMER_CB_IRQSAFE_NO_SOFTIRQ:
 		case HRTIMER_CB_IRQSAFE_NO_RESTART:
+		case HRTIMER_CB_IRQSAFE_PERCPU:
+		case HRTIMER_CB_IRQSAFE_UNLOCKED:
 			cpu_relax();
 			break;
 		default:
